### PR TITLE
[SPARK-19926][PYSPARK] make captured exception from JVM side user friendly

### PR DESCRIPTION
**Pull request by advancedxy**
### What changes were proposed in this pull request? 
The str of `CapaturedException` is now returned by str(self.desc) rather than repr(self.desc), which is more user-friendly. It also handles unicode under python2 specially.

### Why are the changes needed?
This is an improvement, and makes exception more human readable in python side.

### Does this PR introduce any user-facing change?
Before this pr,  select `中文字段` throws exception something likes below:

```
Traceback (most recent call last):
  File "/Users/advancedxy/code_workspace/github/spark/python/pyspark/sql/tests/test_utils.py", line 34, in test_capture_user_friendly_exception
    raise e
AnalysisException: u"cannot resolve '`\u4e2d\u6587\u5b57\u6bb5`' given input columns: []; line 1 pos 7;\n'Project ['\u4e2d\u6587\u5b57\u6bb5]\n+- OneRowRelation\n"
```

after this pr:
```
Traceback (most recent call last):
  File "/Users/advancedxy/code_workspace/github/spark/python/pyspark/sql/tests/test_utils.py", line 34, in test_capture_user_friendly_exception
    raise e
AnalysisException: cannot resolve '`中文字段`' given input columns: []; line 1 pos 7;
'Project ['中文字段]
+- OneRowRelation

```
### How was this patch 
Add a new test to verify unicode are correctly converted and manual checks for thrown exceptions.

This pr's credits should go to @uncleGen and is based on https://github.com/apache/spark/pull/17267

### CODE REVIEW

**Comment by advancedxy** [comment]
Ping @HyukjinKwon and @viirya.

I believe this should be ported before https://github.com/apache/spark/pull/18324, let's check this pr first.

File: python/pyspark/sql/tests/test_utils.py [hunk 1]
```diff
@@ -1,3 +1,4 @@
+# encoding: utf-8
```
Comment (viirya) [inline python/pyspark/sql/tests/test_utils.py]: ` # -*- encoding: utf-8 -*-`?

Comment (advancedxy) [reply python/pyspark/sql/tests/test_utils.py]: Will fix. Do you have any other suggestions?

File: python/pyspark/sql/utils.py [hunk 2]
```diff
@@ -16,6 +16,10 @@
 #
 
 import py4j
+import sys
+
+if sys.version_info.major >= 3:
+    unicode = str
```
Comment (viirya) [inline python/pyspark/sql/utils.py]: do you need this? you only use unicode when version major < 3, isn't?

Comment (advancedxy) [reply python/pyspark/sql/utils.py]: Looks we don't.

I thought `unicode` should be available when running python3. But looks like `isinstance` is not evaluated, and it doesn't require `unicode` to be in the namespace.

**Comment by HyukjinKwon** [comment]
ok to test

File: python/pyspark/sql/tests/test_utils.py (hunk 1 above)
Comment (HyukjinKwon) [reply python/pyspark/sql/tests/test_utils.py]: I think either way is fine (https://www.python.org/dev/peps/pep-0263/#defining-the-encoding)

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
@advancedxy please fix the style. looks good otherwise.

File: python/pyspark/sql/utils.py [hunk 3]
```diff
@@ -25,7 +25,12 @@ def __init__(self, desc, stackTrace, cause=None):
         self.cause = convert_exception(cause) if cause is not None else None
 
     def __str__(self):
-        return repr(self.desc)
+        desc = self.desc
+        # encode unicode instance for python2 for human readable description
+        if sys.version_info.major < 3 and isinstance(desc, unicode):
+            return str(desc.encode('utf-8'))
```
Comment (HyukjinKwon) [inline python/pyspark/sql/utils.py]: Let's remove this once we drop Python 2, which will be right away after Spark 3.0 release.

Comment (advancedxy) [reply python/pyspark/sql/utils.py]: OK.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by tgravescs** [comment]
I'm seeing an error after this went in with unicode:

starting flake8 test...
flake8 checks failed:
./python/pyspark/sql/utils.py:31:60: F821 undefined name 'unicode'
        if sys.version_info.major < 3 and isinstance(desc, unicode):
                                                           ^
1     F821 undefined name 'unicode'

Where is unicode supposed to come from?

**Comment by advancedxy** [comment]
If it’s python2, unicode is built-in.
Otherwise, this part is never evaluated, thus irrelevant.

**Comment by tgravescs** [comment]
hmm, then why am I getting the error? I'm simply running the dev/lint-python script, my default system version is python 3.7.3. it seems to be checking for unicode package still. Its not clear to me which version Jenkins is using. I see it install python 3.5.6 but that is way after the lint script runs

Have you tried this with python3?

**Comment by advancedxy** [comment]
> Have you tried this with python3?

Yes, I ran tests with python3.7 on my local machine, but not `dev/lint-python`.

Looks like flake cannot infer that unicode is never used for python3.  The code is correct in the runtime. 

Seems we don't run lint-python on the Jenkins? @HyukjinKwon can you help confirming that? 

I'd like to address this problem in https://github.com/apache/spark/pull/25847, the fix is simple, just
add the following in the begin:
```
import py4j
import sys

if sys.version_info.major >= 3:
    unicode = str
```

Does that sound reasonable to you @HyukjinKwon ? Or I should submit a follow up pr for this problem? Also cc @viirya

**Comment by tgravescs** [comment]
I would like to see this fixed quickly because dev/lint-python is run by run-tests.py which if you are running with python 3 fails and thus makes my CI fail.  I have to assume the Apache Spark Jenkins is using python 2 at that point.  If #25847 is going to be committed today I'm ok with it, otherwise I think it should be separate pr

**Comment by viirya** [comment]
I'm fine either way. Looks #25847 was approved. It might be merged soon.

**Comment by advancedxy** [comment]
I have updated a new commit to #25847, hope it would be merged soon.

**Comment by viirya** [comment]
I will take a look, If @HyukjinKwon is unable to merge it due to time difference.

**Comment by tgravescs** [comment]
thanks!

**Comment by viirya** [comment]
@tgravescs I have a question on #25847, and I think that @advancedxy can't response in quick due to time difference. I am not sure how many people are affected by this, but in order to unblock it, I am going to create a follow up PR to include the fix.

**Pull request closed without merging**